Add support for IMU Preintegration and discrete-time visual-inertial optimization

B1ueber2y opened the pull request:
Reopening for https://github.com/colmap/colmap/pull/2561 (due to some technical issues during rebase)

B1ueber2y committed:
a small draft (cannot even compile) just to cold start the dev of this feature

B1ueber2y committed:
update c++.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
minor.

B1ueber2y committed:
add checks on finish.

B1ueber2y committed:
hack with mutable and ConvertToDouble.

B1ueber2y committed:
upload pybind.

B1ueber2y committed:
fix format.

B1ueber2y committed:
merge with main.

B1ueber2y committed:
change imumeasurements to priority list.

B1ueber2y committed:
fix format.

B1ueber2y committed:
support optimization on gravity.

B1ueber2y committed:
fix format.

B1ueber2y committed:
add repr for ImuMeasurement.

B1ueber2y committed:
update.

B1ueber2y committed:
add setters for imu state variables.

B1ueber2y committed:
velocities should be multiplied by the scale as well.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
minor.

B1ueber2y committed:
Update src/colmap/geometry/pose.cc

Co-authored-by: Johannes Schönberger <jsch@demuc.de>

B1ueber2y committed:
fix last commit and format.

B1ueber2y committed:
update.

B1ueber2y committed:
minor update.

B1ueber2y committed:
update.

B1ueber2y committed:
not need to sort when initialized from another ImuMeasurement.

B1ueber2y committed:
add a todo.

B1ueber2y committed:
change convention and fix transformation.

B1ueber2y committed:
update binding.

B1ueber2y committed:
add some serialization.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
add support for rectification matrix.

B1ueber2y committed:
reorder field.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
update.

B1ueber2y committed:
svd and improve conditioning.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
update. align conventions.

B1ueber2y committed:
add gyro check at reintegration.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
add a test in python with mps data.

B1ueber2y committed:
update.

B1ueber2y committed:
add more options in the example on what to optimize.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
update.

B1ueber2y committed:
formatting.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
add impl.

B1ueber2y committed:
add pybind

B1ueber2y committed:
add tests.

B1ueber2y committed:
update.

B1ueber2y committed:
update.

B1ueber2y committed:
refactor. separate the manifold file. Add PositiveExponentialManifold

B1ueber2y committed:
formatting.

B1ueber2y committed:
update.

B1ueber2y committed:
add scale_ptr for sim3d pybind.

B1ueber2y committed:
fix.

B1ueber2y committed:
fix.

B1ueber2y committed:
update.

B1ueber2y committed:
fix binding for sim3d.scale.

B1ueber2y committed:
update.

B1ueber2y committed:
formatting.

B1ueber2y committed:
test pyceres.

B1ueber2y committed:
test.

B1ueber2y committed:
update.

B1ueber2y committed:
move pyceres checker to pycolmap/helpers.h + minor.

B1ueber2y committed:
Merge branch 'features/alignment_cost' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
fix covariane propagation.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
update covariance propagation.

robinlindh commented:
Hello! Very nice work! Would be very exciting to get this kind visual-inertial optimization into colmap.

I'm right now trying this out, as it fits a refinement pipeline I work on. I currently use only visual feature matching, so I wanted to try picking the bundle adjustment parts from this PR for adding IMU too.

So far I got it to run similar to in your example.py, but I'm getting a very large cost (some trillions initially and then decreasing to billions after like 50 iterations of ceres solver). Might be that I'm passing IMU readings in the wrong coordinate convention somehow.

Which coordinate system convention, units etc does your IMU code expect? It was a bit hard to judge just from the example code (I did download that np pickle which gives some insights but not sure still). It seems that downloaded file in the example has also done some kind of pre-processing already?

Right now I'm capturing raw Accelerometer and Gyro from ARKit for my dataset, but flipping axes a bit to fit the colmap coordinate system conventions. And some interpolation logic to match timestamps between the gyro and accelerometer.
I was also a bit unsure whether gyro values can be used as is, or if the values need to somehow be converted into "angle axis" which your code seems to expect.

Thankful for any help, and happy to report back here with my results later.

B1ueber2y commented:
Hey. Thanks for your interest.

> So far I got it to run similar to in your example.py, but I'm getting a very large cost (some trillions initially and then decreasing to billions after like 50 iterations of ceres solver). Might be that I'm passing IMU readings in the wrong coordinate convention somehow.

The noise parameters in the example is picked from the aria doc and are known to be with significant underestimation, which will result in a large cost in general. Using the right noise parameters corresponding to your IMU is crucial for incorporating the preintegration cost into your pipeline.  

> Right now I'm capturing raw Accelerometer and Gyro from ARKit for my dataset, but flipping axes a bit to fit the colmap coordinate system conventions. And some interpolation logic to match timestamps between the gyro and accelerometer. I was also a bit unsure whether gyro values can be used as is, or if the values need to somehow be converted into "angle axis" which your code seems to expect.

The unit of the gyro values in the example is rad/s. 

This feature will likely not be merged very soon. But feel free to give more valuable feedback here. Thanks! : )

robinlindh commented:
Thanks for the reply!
I managed to find out the correct coordinate conversion I think between ARKit raw IMU capture and what you use here.
In case it's of help to anyone else, I ended up just swapping the order of X and Y for both the gyro and IMU.
That on its own brought the cost down a lot and seems to give better results too.

As for the IMU noise and random walk, I just now changed the values to more appropriate ones for iPhone, and indeed the costs are much more reasonable now, similar order of magnitude as my total reprojection cost from visual features.

> This feature will likely not be merged very soon. But feel free to give more valuable feedback here. Thanks! : )

No worries at all! I know the PR is still in draft stage too and that there was more work you wanted to do for full integration. For me I can already use it and hopefully just more feedback here can help others and eventually help this PR to get merged :)

Cheers!

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
update hyperpameters from the web update.

ahojnnes committed:
Merge branch 'main' into features/imu

ahojnnes committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
update.

B1ueber2y committed:
merge with main.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
normalize quaternion.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge and resolve conflicts.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge and resolve conflicts.

B1ueber2y committed:
fix linting errors.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge from main and resolve conflicts.

B1ueber2y committed:
fir merge conflicts.

B1ueber2y committed:
merge with main and resolve conflicts.

B1ueber2y committed:
support positional covariance from integration noise.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge and resolve conflicts

B1ueber2y committed:
Merge branch 'main' into features/imu

davidwatkins-bdai commented:
@B1ueber2y what is the status on this PR? Is there ongoing work that needs to be done in order to finalize this? The example currently does not run as is with your most recent changes.

mvictora-bdai commented:
I think the main problem with the current example is that it doesn't use the changes introduced in https://github.com/colmap/colmap/pull/2896/files . However, I have a change that does work. I don't have permissions to push my branch but I basically just changed one function:

```python
def solve_bundle_adjustment(
    reconstruction, ba_options, ba_config, preintegrated_measurements, variables
):
    bundle_adjuster = pycolmap.create_default_bundle_adjuster(ba_options, ba_config, reconstruction)

    summary = bundle_adjuster.solve()

    print(summary.BriefReport())
    return summary
```

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
adapt to the new colmap update. fix quaternion convention.

B1ueber2y committed:
format

B1ueber2y commented:
> @B1ueber2y what is the status on this PR? Is there ongoing work that needs to be done in order to finalize this? The example currently does not run as is with your most recent changes.

Hellooo. Thanks for the interest. This has been fixed just now. This PR likely wont be merged very soon since we plan to only support IMU after supporting rig data structure in COLMAP https://github.com/colmap/colmap/pull/2698. I will try to find some time to push this in the following weeks.

mvictora-bdai commented:
> Hellooo. Thanks for the interest. The example has been fixed just now. This PR likely wont be merged very soon since we plan to only support IMU preintegration after supporting rig data structure in COLMAP #2698. I will try to find some time to push this in the following weeks.

Asking for a bit of clarification here -- does the code work as-is (at least if you're able to adapt your data to fit to something like the example)? You're waiting for rig data structure work to be merged, and then you will rework this code to be built on top of a multi-sensor rig data structure before merging it to main? Will this new data structure help with being able to achieve better state estimation?

B1ueber2y commented:
> > Hellooo. Thanks for the interest. The example has been fixed just now. This PR likely wont be merged very soon since we plan to only support IMU preintegration after supporting rig data structure in COLMAP #2698. I will try to find some time to push this in the following weeks.
> 
> Asking for a bit of clarification here -- does the code work as-is (at least if you're able to adapt your data to fit to something like the example)? You're waiting for rig data structure work to be merged, and then you will rework this code to be built on top of a multi-sensor rig data structure before merging it to main? Will this new data structure help with being able to achieve better state estimation?

Yes your understanding is correct. The feature is completed and we are waiting for the rig support to be completed. There is no point to support IMU preintegration in COLMAP without rig support, since this is only useful for visual-inertial optimization. Thanks!

mvictora-bdai commented:
I want to probe a bit deeper on what rig support will provide -- right now, I want to use IMU data from a GoPro to improve localization I achieve using glomap. However, to do that I need to:

1. Interpolate/align my IMU timestamps with my RGB frames
2. Handle the transform from my IMU sensor to my camera

Will https://github.com/colmap/colmap/pull/2698 help me handle those needs?

B1ueber2y commented:
> I want to probe a bit deeper on what rig support will provide -- right now, I want to use IMU data from a GoPro to improve localization I achieve using glomap. However, to do that I need to:
> 
>     1. Interpolate/align my IMU timestamps with my RGB frames
> 
>     2. Handle the transform from my IMU sensor to my camera
> 
> 
> Will #2698 help me handle those needs?

The rig support can help you handle the transformation from your IMU sensor to the camera. However, we dont plan to support synchronization optimization in the very near future, so you will have to provide synchronization yourself. If the timestamps of your IMU measurements and RGB images are well synchronized we could help handle it in the visual-inertial optimization.

B1ueber2y committed:
merge from main and resolve conflicts

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
use features from pybind 3.0.0.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
tmp

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
revert tmp.

sarlinpe commented on src/colmap/estimators/imu_preintegration.h:
```diff
@@ -0,0 +1,304 @@
+  bool HasFinished() const;
+
+  // Reintegrate
+  bool CheckReintegrate(const Eigen::Vector6d& biases) const;
```
ShouldReintegrate might be more informative.

sarlinpe commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,412 @@
+  Reintegrate();
+}
+
+// data interfaces
```
Should all of these be inlined?

sarlinpe reviewed (commented):
I haven't yet checked the maths 😬

B1ueber2y committed:
merge upstream and resolve merge conflicts.

B1ueber2y committed:
cr.

B1ueber2y committed:
cr.

B1ueber2y committed:
use int64_t instead of double for timestamps in nanoseconds.

B1ueber2y committed:
add back python examples with typing.

B1ueber2y committed:
format.

B1ueber2y committed:
consolidate with internal prototyping.

B1ueber2y committed:
fix python scripts.

B1ueber2y committed:
single-pose parameter block.

B1ueber2y committed:
Use ceres callback to reintegrate to avoid mutation.

B1ueber2y committed:
rectification matrix goes into calibration.

B1ueber2y committed:
Clean rework on two separate cost functions.

B1ueber2y committed:
add tests and fix.

B1ueber2y committed:
rename functions to functors. format.

B1ueber2y committed:
align header changes.

B1ueber2y committed:
minor.

B1ueber2y committed:
improve comments and fix stubs.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
further consolidation with new Rigid3d design.

B1ueber2y committed:
minor renaming.

B1ueber2y committed:
rename acc to accel.

B1ueber2y committed:
fix python convention and typing.

B1ueber2y committed:
in preparation for more advanced integrators.

B1ueber2y committed:
add golden reference.

B1ueber2y committed:
switch to left convention implementation.

B1ueber2y committed:
add draft RK4 impl.

B1ueber2y committed:
stabilize

B1ueber2y committed:
correct convention and stabilize tests.

B1ueber2y committed:
minor improvements on comments.

B1ueber2y committed:
format.

B1ueber2y committed:
cleanup.

B1ueber2y committed:
add max condition number into option.

B1ueber2y committed:
analytical jacobians

B1ueber2y committed:
rename to SetLinearizationBiases

B1ueber2y committed:
format.

B1ueber2y committed:
analytical jacobian and pybind.

B1ueber2y committed:
clang tidy.

B1ueber2y committed:
dont set max_condition_number by default.

B1ueber2y committed:
do not clamp to zero.

ahojnnes commented on src/colmap/scene/imu.h:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  camera_t imu_id = kInvalidCameraId;
+  image_t image_id = kInvalidImageId;
```
Should these measurements more generally be associated as sensor_t and data_t instead of camera/image?

ahojnnes commented on src/colmap/estimators/cost_functions/quaternion_utils_test.cc:
```diff
@@ -0,0 +1,138 @@
+
+TEST(QuaternionRotatePointWithJac, Nominal) {
+  SetPRNGSeed(42);
+  const double eps = 1e-7;
```
Same elsewhere.

ahojnnes reviewed (commented):
Looks fantastic. Made a first (very superficial) pass and need to carve out some dedicated time to look at this in detail. Happy holidays in the meanwhile.

B1ueber2y committed:
drop imu_id and image_id from ImuState.

B1ueber2y committed:
reorder to make sure that gyro always comes before accel.

B1ueber2y committed:
cr.

B1ueber2y committed:
cr.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
improve comments on integration noise density and fix its usage in RK4.

B1ueber2y committed:
Merge branch 'tmp/backup' into features/imu

B1ueber2y committed:
improve comment for midpoint vs. RK4.

B1ueber2y commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,587 @@
+  data_.covariance.block<3, 3>(12, 12) += Eigen::Matrix3d::Identity() * vars_ba;
+}
+
+void ImuPreintegrator::IntegrateRK4(const Eigen::Vector3d& accel_true,
```
I did not understand this comment. Could you help clarify? Thanks.

B1ueber2y committed:
move reintegrate related options into callback class.

B1ueber2y committed:
typing fix.

B1ueber2y committed:
cr.

B1ueber2y committed:
revive ImuMeasurements class and add tests.

B1ueber2y committed:
format.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
fix move.

TannerGilbert committed:
Fixed vi_optimization.py

- Updated data to have time format as an int
- Use `pyceres.SolverOptions` as `ceres` bindings in `pycolmap` doesn't have `.callbacks`

B1ueber2y committed:
Merge pull request #2 from TannerGilbert/patch-1

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge upstream and resolve conflicts.

B1ueber2y committed:
change convention to AFromB.

B1ueber2y committed:
merge upstream and resolve conflicts.

B1ueber2y committed:
merge upstream and resolve conflicts.

B1ueber2y committed:
merge upstream and resolve conflicts.

B1ueber2y committed:
move imu preintegration cost upwards.

B1ueber2y committed:
specify project aria gen 1 for the example

B1ueber2y committed:
misc update.

B1ueber2y committed:
add a small comment.

B1ueber2y commented on src/colmap/estimators/imu_preintegration.h:
```diff
@@ -0,0 +1,243 @@
+                    const Eigen::Vector3d& gyro_true,
+                    double dt,
+                    double accel_noise_density,
+                    double gyro_noise_density);
```
We only have two methods and I do not expect us to support more in the future. Both methods do not have any persistent method-specific config/state. So I would vote for keeping it as it currently is.

ahojnnes replied:
The benefit of having an abstract interface is that we can have parameterized tests or mock/fake the implementation when the pre-integrator is part of another pipeline. Unless it's difficult to implement (and probably Claude can one-shot the change?) or there is some other conceptual blocker, I'd suggest to make the change.

B1ueber2y replied:
This makes great sense to me. I have added the abstraction and changed to parameterized tests.

B1ueber2y commented on src/colmap/estimators/imu_preintegration_cost.h:
```diff
@@ -0,0 +1,856 @@
+  const PreintegratedImuData* data_;
+};
+
+// Analytical-Jacobian version of VisualCentricImuPreintegrationCostFunctor.
```
The analytical jacobian on the cost functors itself have large speed up factors.


| case | AutoDiff | Analytical | speedup |
  |----------------|----------|------------|---------|
  | body-centric | ~18.4 µs | ~2.29 µs | ~8× |
  | visual-centric | ~35.2 µs | ~3.39 µs | ~10× |

B1ueber2y committed:
add a runtime benchmark on Evaluate.

B1ueber2y committed:
format.

B1ueber2y committed:
format.

B1ueber2y committed:
remove an unexpectedly committed file.

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+// https://github.com/uzh-rpg/rpg_svo_pro_open/blob/master/svo_common/include/svo/common/imu_calibration.h
+// Default parameters are for ADIS16448 IMU.
+struct ImuCalibration {
+  /// Gyro noise density (sigma). [rad/s*1/sqrt(Hz)]
```
nit: Applies also to other comments below.
```suggestion
  // Gyro noise density (sigma). [rad/s*1/sqrt(Hz)]
```

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+  double imu_rate = 20.0;
+
+  /// Rectification matrices correcting axis misalignment and scale.
+  /// m_true = M^{-1}(m - b). Identity if data is already rectified.
```
What is b and m here?

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+  /// m_true = M^{-1}(m - b). Identity if data is already rectified.
+  /// TODO: Support online calibration by making these optimizable
+  /// as parameter blocks in the IMU cost function.
+  Eigen::Matrix3d gyro_rectification = Eigen::Matrix3d::Identity();
```
Can we be more explicit about the transformation? For example, `gyro_true = true_from_measured * gyro_measured` or similar. Just making these up, as I have not read the whole code and starting here with the review.

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+
+struct ImuMeasurement {
+  timestamp_t timestamp = kInvalidTimestamp;  // [nanoseconds]
+  Eigen::Vector3d gyro = Eigen::Vector3d::Zero();
+  Eigen::Vector3d accel = Eigen::Vector3d::Zero();
```
Is 0 a good default or use quiet_NaN similar to pose prior?

B1ueber2y replied:
These two values are always available for every Imu measurement, so is never optional as in the case of pose prior. So I personally think having the default as NaN is not as good as zero.

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+    if (m.timestamp == measurements_.back().timestamp) {
+      throw std::invalid_argument("Duplicate timestamp in ImuMeasurements: " +
+                                  std::to_string(m.timestamp));
+    }
```
Wouldn't this already be handled by the below check?

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+                });
+      ThrowIfHasDuplicates(measurements_);
+    } else {
+      for (auto it = ms.begin(); it != ms.end(); ++it) Insert(*it);
```
This loop has quite bad performance. Wouldn't it be more efficient to just append all measurements at the end of the vector and then sort + ThrowIfDuplicates? Alternatively, take the vector by copy, then sort and use InsertSorted?

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+    Insert(ms);
+  }
+
+  void Insert(const ImuMeasurement& m) {
```
Do we really need this interface? It has bad complexity.

B1ueber2y replied:
I can see its usability in the stream setup. I would vote for keeping it. I updated the other suboptimal insertion methods.

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+                         const ImuMeasurement& measurement);
+
+// Sorted list of IMU measurements ordered by timestamp.
+class ImuMeasurements {
```
* The methods in the header of this class probably all need inlining otherwise we'll end up with duplicate symbols?
* I'd suggest moving most of the heavier methods into the .cc file.

ahojnnes commented on src/colmap/sensor/imu.h:
```diff
@@ -0,0 +1,234 @@
+  /// Norm of the Gravitational acceleration. [m/s^2]
+  double gravity_magnitude = 9.81007;
+
+  /// Expected IMU Rate [1/s]
```
nit
```suggestion
  /// Expected IMU rate. [1/s]
```

ahojnnes commented on src/colmap/sensor/imu.cc:
```diff
@@ -0,0 +1,78 @@
+
+namespace colmap {
+
+ImuMeasurements ImuMeasurements::ExtractMeasurementsContainEdge(
```
nit: Maybe better?
```suggestion
ImuMeasurements ImuMeasurements::ExtractMeasurementsInRange(
```

ahojnnes commented on src/colmap/sensor/imu.cc:
```diff
@@ -0,0 +1,78 @@
+  auto cmp = [](const ImuMeasurement& m1, const ImuMeasurement& m2) {
+    return m1.timestamp < m2.timestamp;
+  };
+  ImuMeasurement dummy;
```
```suggestion
  ImuMeasurement range;
```

ahojnnes commented on src/colmap/sensor/imu_test.cc:
```diff
@@ -0,0 +1,224 @@
+
+const Eigen::Vector3d kZero = Eigen::Vector3d::Zero();
+
+ImuMeasurement M(timestamp_t t) { return ImuMeasurement(t, kZero, kZero); }
```
CreateMeasurement?

ahojnnes commented on src/colmap/sensor/imu_test.cc:
```diff
@@ -0,0 +1,224 @@
+
+ImuMeasurement M(timestamp_t t) { return ImuMeasurement(t, kZero, kZero); }
+
+TEST(ImuMeasurements, Nominal) {
```
Suggest decomposing these into dedicated tests for each "ms".

ahojnnes commented on src/pycolmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,100 @@
+#include "colmap/estimators/imu_preintegration.h"
```
Let's cover these new bindings in our bindings tests.

ahojnnes commented on src/colmap/scene/imu.h:
```diff
@@ -0,0 +1,111 @@
+namespace colmap {
+
+// An Imu class storing the sensor information and a linked visual camera.
+// TODO: Integrate with the Rig abstraction (sensor/rig.h) by making IMU a
```
Why not now? Feels like we will otherwise create a lot of technical debt?

B1ueber2y replied:
This PR focuses on discrete-time VI optimization with IMU preintegration, which is a cost level update for post optimization that is not coupled into the sfm pipeline. To achieve this we do not need to necessarily benefit from the rig abstraction for now. 

Integrating it with the rig abstraction requires some additional efforts on updating data structure in the reconstruction and incremental pipeline to include ImuMeasurement and ImuState (still need to think of how to make it memory efficient), which would make this already gigantic PR even harder to review. So I suggest we keep this TODO out of the scope of this PR.

ahojnnes commented on src/colmap/scene/imu.h:
```diff
@@ -0,0 +1,111 @@
+  }
+};
+
+inline std::ostream& operator<<(std::ostream& stream, const Imu& imu) {
```
Move to .cc

ahojnnes commented on src/colmap/scene/imu.h:
```diff
@@ -0,0 +1,111 @@
+  return stream;
+}
+
+inline std::ostream& operator<<(std::ostream& stream, const ImuState& state) {
```
Move to .cc

ahojnnes commented on src/colmap/scene/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ COLMAP_ADD_LIBRARY(
         camera.h camera.cc
         frame.h frame.cc
         image.h image.cc
+        imu.h
```
Add tests.

ahojnnes commented on src/colmap/sensor/imu_test.cc:
```diff
@@ -0,0 +1,224 @@
+
+#include <gtest/gtest.h>
+
+namespace colmap {
```
Add test for ostream. See all other structs/classes with equivalent tests.

ahojnnes commented on src/colmap/estimators/imu_preintegration.h:
```diff
@@ -0,0 +1,259 @@
+// Right-perturbation model for bias Jacobians:
+//   delta_R(bg + dbg) ≈ delta_R * Exp(dR_dbg * dbg)
+//   The bias correction is always right-multiplied in the cost function.
+struct PreintegratedImuData {
```
nit: Maybe PreintegratedImuState ?

B1ueber2y replied:
State is already taken by ImuState and it means something different. After preintegration it s not really a state anymore, so I would prefer the original one "data" out of the two.

ahojnnes commented on src/colmap/estimators/imu_preintegration.h:
```diff
@@ -0,0 +1,259 @@
+
+  // IMU Calibration.
+  const ImuCalibration calib_;
+  Eigen::Matrix3d accel_rect_mat_inv_ = Eigen::Matrix3d::Identity();
```
Same comment as in sensor/imu.h about using more explicit notation for the transformation. In this case, maybe `measured_from_true` ?

ahojnnes commented on src/colmap/estimators/imu_preintegration.h:
```diff
@@ -0,0 +1,259 @@
+  double reintegrate_vel_norm_thres = 1e-4;
+};
+
+class ImuReintegrationCallback : public ceres::IterationCallback {
```
Should this go into another file (e.g., imu_reintragration.h) or follow-up PR?

B1ueber2y replied:
It belongs to this PR because we use it in the python example. And I dont think its worth splitting to another file. 

To be clear on what this was trying to achieve, in the imu preintegration convention (e.g., in VI.C of Forster et al. https://rpg.ifi.uzh.ch/docs/TRO16_forster.pdf), the bias was handled as constant with a linear jacobian approximation, which is oftentimes accurate enough. However, if the bias drifts too much its common practice to reintegrate to get an updated linearized jacobian, which is in practice quite crucial. Here we achieve the same behavior with a ceres callback.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  covariance = (covariance + covariance.transpose()) / 2.0;
+
+  // Eigendecomposition for robust sqrt-information.
+  Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, 15, 15>> saes(covariance);
```
const

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  Reset();
+}
+
+void ImuPreintegrator::Reset() {
```
Maybe instead just do `data_ = ImuPreintegrationData()` ?

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  // Integration step.
+  // dR = Exp(-omega * dt): body_from_world evolves as R_BW_{k+1} = Exp(-w*dt) *
+  // R_BW_k.
+  Eigen::Quaterniond dq = QuaternionFromAngleAxis(-gyro_true * dt);
```
nit: some of the variables here can benefit from const's

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  data_.covariance = A * data_.covariance * A.transpose();
+
+  // Step 2: add noise.
+  double vars_v = pow(accel_noise_density, 2) * dt;
```
pow() is slow, replace with a * a.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  double vars_ba = pow(calib_.bias_accel_random_walk_sigma, 2) * dt;
+  double vars_bg = pow(calib_.bias_gyro_random_walk_sigma, 2) * dt;
+  data_.covariance.block<3, 3>(0, 0) +=
+      Eigen::Matrix3d::Identity() * vars_omega;
```
I think you can do `.diagonal() += vars_omega`. Same for below.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+
+  // Integration matrices for position (H_p) and velocity (H_v).
+  // Closed-form integral coefficients (same signs as Eckenhoff IJRR 2018).
+  const Eigen::Matrix3d alpha_arg = (dt2 / 2.0) * I3 + f_1 * w_x + f_2 * w_x_2;
```
nit: `0.5 * dt2` is more efficient.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  // dR = Exp(-omega * dt): body_from_world evolves as R_BW_{k+1} = Exp(-w*dt) *
+  // R_BW_k.
+  Eigen::Quaterniond dq = QuaternionFromAngleAxis(-gyro_true * dt);
+  Eigen::Matrix3d Rs = data_.delta_R.toRotationMatrix();
```
Should we store delta_R as a rotation matrix? It seems like we convert back and forth without ever needing the quaternion during integration?

B1ueber2y replied:
quaternion is used in the cost functions and bias multiplication, where we need the performance the most. So keeping quaternion over matrix makes sense to me.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  // Based on Eckenhoff et al. "Closed-form Preintegration Methods for
+  // Graph-based Visual-Inertial Navigation", IJRR 2018 (left convention).
+
+  const Eigen::Matrix3d I3 = Eigen::Matrix3d::Identity();
```
Can be a static constexpr in anonymous namespace?

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  const Eigen::Vector3d e_3(0, 0, 1);
+  const Eigen::Matrix3d e_1x = CrossProductMatrix(e_1);
+  const Eigen::Matrix3d e_2x = CrossProductMatrix(e_2);
+  const Eigen::Matrix3d e_3x = CrossProductMatrix(e_3);
```
Could statically precompute these.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  }
+
+  const Eigen::Matrix3d* d_R_bw[3] = {&d_R_bw_1, &d_R_bw_2, &d_R_bw_3};
+  const Eigen::Matrix3d* e_kx[3] = {&e_1x, &e_2x, &e_3x};
```
Alternatively could store these as directly in an std::array above instead of the extra pointer indirection.

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  // Continuous-time noise matrix Q_c (12x12):
+  // [gyro_noise(3), gyro_walk(3), accel_noise(3), accel_walk(3)].
+  Eigen::Matrix<double, 12, 12> Q_c = Eigen::Matrix<double, 12, 12>::Zero();
+  Q_c.block<3, 3>(0, 0) = I3 * (gyro_noise_density * gyro_noise_density);
```
Can use .diagonal() here as well?

ahojnnes commented on src/colmap/estimators/imu_preintegration.cc:
```diff
@@ -0,0 +1,589 @@
+  IntegrateOneMeasurement(last_measurement, m);
+}
+
+void ImuPreintegrator::FeedImu(const ImuMeasurements& ms) {
```
nit: Integrate() ?

ahojnnes reviewed (commented):
Thanks, looks great already. Left an initial round of comments.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Update src/colmap/sensor/imu.h

Co-authored-by: Johannes Schönberger <jsch@demuc.de>

B1ueber2y committed:
Update src/colmap/sensor/imu.h

Co-authored-by: Johannes Schönberger <jsch@demuc.de>

B1ueber2y committed:
cr.

B1ueber2y committed:
change integrator test to parameterized one.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
Merge branch 'tmp/backup' into features/imu

B1ueber2y committed:
cr.

B1ueber2y committed:
add some nominal python tests.

B1ueber2y committed:
rename FeedImu to Integrate and update pybind.

B1ueber2y committed:
document affine translation and rename to true_from_measured.

B1ueber2y committed:
strip INVLAID_TIMESTAMP binding due to a separate PR.

B1ueber2y committed:
move heavy methods of ImuMeasurement to cc.

B1ueber2y commented:
Thanks for the detailed review! All comments addressed for this round : )

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
merge upstream and resolve conflicts.

B1ueber2y committed:
Merge branch 'main' into features/imu

B1ueber2y committed:
fix the year of the Eckenhoff IJRR reference.
